File: wcserver/delphi/examples/wcYell/wcyell.pas
{$APPTYPE CONSOLE}
{$H+,O+,A-}
Uses wcserver,
     Windows,
     SysUtils;

type
  TPageMessage = record
    From  : Array[0..27] of char;
    Text  : Array[1..3] of Array[0..79] of char;
    InviteToChat : LongInt;
  end;

Var
   pmsg  : TPageMessage;

(*
Procedure ShowServerLoginError(const err : integer);
var note : string;
  begin
    MessageBeep(0);
    note := 'Check to see if WCSERVER is loaded';
    case err of
      WC_INCORRECT_PASSWORD:
        note := 'Check System Password';
    end;

    Writeln('Wilcat! Server Login Failure');
    writeln(Format('Server Error (%8X) %s'#13#13,[err,GetWildcatErrorStr(err)]));
    writeln(Note);
  end;
*)

Function ParamLineRaw:String;
(* return the entire command line minus program name *)
begin
  ParamLineRaw := '';
  if pos(' ',CmdLine) = 0 then exit;
  ParamLineRaw := Copy(CmdLine,Pos(' ',Cmdline)+1,length(cmdline));
end;


Begin
  if ParamCount=0 then
     begin
       writeln('wcYell v8.0 (c) copyright 1998-2019 Santronics Software');
       writeln('syntax: wcYell <broadcast message>');
       writeln;
       writeln('wcYell will broadcast a message to all nodes. The maximum');
       writeln('length of the message is 78 characters');
       writeln;
       writeln('example: wcYell Please Log off! System Maintenance begins in 1 min');
       exit;
     end;

  if not WildcatServerConnect(0) then
     begin
       //ShowServerLoginError(GetLastError());
       exit;
     end;

  try
     if not WildcatServerCreateContext() then
        begin
          //ShowServerLoginError(GetLastError());
          writeln('Error Wildcat! Create Context Error: ',GetLastError());
          exit;
        end;

     LoginSystem();

     fillchar(pmsg,sizeof(pmsg),#0);
     pmsg.From := 'SYSOP';
     strPcopy(pmsg.Text[1],copy(ParamLineRaw,1,78));

     WriteChannel(wcserver.OpenChannel('System.Page'),
                   0,
                   0,
                   @pmsg,
                   sizeof(pmsg));
  finally
     WildcatServerDeleteContext();
  end;

End.

File: wcserver/cpp/examples/channels/wcyell.cpp
// File: wcYell.cpp

#include <stdio.h>
#include <stdio.h>
#include <windows.h>
#include <wcserver.h>
#include <wcversion.h>
#include <wclinker.h>


void Yell(char argc, char *argv[])
{
   TPageMessage page = {"Sysop",""};
   char sz[sizeof(page.message[0])] = {0};
   for (int i=1; i<argc; i++) {
      char *p = argv[i];
      if (strlen(sz)+strlen(p) < sizeof(sz)) {
         if (i > 1 && (sz[strlen(sz)-1] != ' ')) strncat(sz," ",sizeof(sz)-1);
         strncat(sz,p, sizeof(sz)-1);
      } else {
         strncat(sz,p, sizeof(sz)-1);
         break;
      }
   }
   strncpy(page.message[0], sz, sizeof(page.message));
   WriteChannel(OpenChannel("System.Page"),0,0,&page,sizeof(page));
}

int main(char argc, char *argv[])
{
   if (argc < 2) {
       printf("wcYell %s %s\n", WC_VERSION_STR, WC_COPYRIGHT_LONG);
       printf("syntax: wcYell <broadcast message>\n");
       printf("\n");
       printf("wcYell will broadcast a message to all nodes. The maximum\n");
       printf("length of the message is 78 characters\n");
       printf("\n");
       printf("example: wcYell Please Log off! System Maintenance begins in 1 min\n");
       return 1;
   }

   if (!WildcatServerConnect(NULL)) return 1;
   if (!WildcatServerCreateContext()) return 1;

   __try {
     if (!LoginSystem()) return 1;
     Yell(argc, argv);
   } __finally {
     WildcatServerDeleteContext();
   }

   return 0;
}
